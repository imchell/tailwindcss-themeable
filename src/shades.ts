import { ColorHex, ColorShadesComputed, lightFactorDefault, saturationFactorDefault, shadeStops } from '.'
import { hex2HSL, hsl2Hex } from './utils'

/**
 * Generate 10 shades (from 50 to 900) from one color
 * Thanks to: https://github.com/nickgraffis/tailwind-color-generator
 */
export const color2Shades = (color: ColorHex, saturationFactor = saturationFactorDefault, lightFactor = lightFactorDefault, isDark = false) => {
  // find the best shade index for this color, if this color is too light, it will be closer to `50`, otherwise it will be closer to `900` if too dark
  const [h, s, l] = hex2HSL(color)

  const lightness = [97]
  while (lightness.length < shadeStops.length) {
    lightness.push(lightness[lightness.length - 1] - lightFactor)
  }
  const closestLightness = lightness.reduce((prev, curr) => {
    return (Math.abs(curr - l) < Math.abs(prev - l) ? curr : prev)
  })
  const closestShadeIndex = lightness.indexOf(closestLightness)

  const ans: ColorShadesComputed = {
    DEFAULT: color
  } as any
  // For darker shades
  for (let i = closestShadeIndex; i < shadeStops.length; i++) {
    const step = i - closestShadeIndex
    ans[shadeStops[i]] = hsl2Hex([h, s - saturationFactor * step, l - lightFactor * step])
  }
  // For lighter shades
  for (let i = 0; i < closestShadeIndex; i++) {
    const step = closestShadeIndex - i
    ans[shadeStops[i]] = hsl2Hex([h, s + saturationFactor * step, l + lightFactor * step])
  }

  // reverse the shades if it is a dark theme, see the type docs of Theme['isDark']
  if (isDark) {
    for (let i = 0, j = shadeStops.length - 1; i < j; i++, j--) {
      [ans[shadeStops[i]], ans[shadeStops[j]]] = [ans[shadeStops[j]], ans[shadeStops[i]]]
    }
  }

  return ans
}
