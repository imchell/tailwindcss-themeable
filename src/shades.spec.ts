import { color2Shades } from './shades'
import { themeDracula } from '.'

describe('shades', () => {
  it('should generate shades from color correctly', () => {
    for (const color of Object.values(themeDracula.palette)) {
      expect(color2Shades(color)).toMatchSnapshot()
    }
  })
})
