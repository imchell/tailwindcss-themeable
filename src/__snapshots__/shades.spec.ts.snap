// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shades should generate shades from color correctly 1`] = `
Object {
  "100": "#b5bad6",
  "200": "#9ea4c8",
  "300": "#888fb8",
  "400": "#727aa8",
  "50": "#cdd0e4",
  "500": "#5f6795",
  "600": "#53597c",
  "700": "#454a64",
  "800": "#373a4d",
  "900": "#282a36",
  "DEFAULT": "#282A36",
}
`;

exports[`shades should generate shades from color correctly 2`] = `
Object {
  "100": "#eaeada",
  "200": "#dcdcc3",
  "300": "#cdcdac",
  "400": "#bdbd96",
  "50": "#f8f8f2",
  "500": "#adad81",
  "600": "#9c9c6c",
  "700": "#85855d",
  "800": "#6d6d4f",
  "900": "#565641",
  "DEFAULT": "#F8F8F2",
}
`;

exports[`shades should generate shades from color correctly 3`] = `
Object {
  "100": "#dfe1ed",
  "200": "#c7cadf",
  "300": "#b1b5d0",
  "400": "#9ba0c0",
  "50": "#f7f7fb",
  "500": "#858bb0",
  "600": "#71779f",
  "700": "#606689",
  "800": "#525771",
  "900": "#44475a",
  "DEFAULT": "#44475A",
}
`;

exports[`shades should generate shades from color correctly 4`] = `
Object {
  "100": "#d7dcec",
  "200": "#bec6df",
  "300": "#a6b0d1",
  "400": "#8f9bc3",
  "50": "#f0f2f8",
  "500": "#7886b4",
  "600": "#6272a4",
  "700": "#54628c",
  "800": "#485273",
  "900": "#3b425a",
  "DEFAULT": "#6272A4",
}
`;

exports[`shades should generate shades from color correctly 5`] = `
Object {
  "100": "#d4f8ff",
  "200": "#aff0fe",
  "300": "#8be9fd",
  "400": "#67e1fb",
  "50": "#fafeff",
  "500": "#44d9f8",
  "600": "#22d0f5",
  "700": "#0dbfe5",
  "800": "#0ca0bf",
  "900": "#0c819a",
  "DEFAULT": "#8BE9FD",
}
`;

exports[`shades should generate shades from color correctly 6`] = `
Object {
  "100": "#bcffcd",
  "200": "#97feb1",
  "300": "#73fc96",
  "400": "#50fa7b",
  "50": "#e2ffe9",
  "500": "#2df760",
  "600": "#0cf346",
  "700": "#0ccd3d",
  "800": "#0ba833",
  "900": "#0a8329",
  "DEFAULT": "#50FA7B",
}
`;

exports[`shades should generate shades from color correctly 7`] = `
Object {
  "100": "#ffdcb7",
  "200": "#ffca92",
  "300": "#ffb86c",
  "400": "#fda648",
  "50": "#ffefdd",
  "500": "#fb9325",
  "600": "#f38107",
  "700": "#cd6d08",
  "800": "#a75a08",
  "900": "#824707",
  "DEFAULT": "#FFB86C",
}
`;

exports[`shades should generate shades from color correctly 8`] = `
Object {
  "100": "#ffc4e6",
  "200": "#ff9fd6",
  "300": "#ff79c6",
  "400": "#fd55b6",
  "50": "#ffeaf6",
  "500": "#fb31a5",
  "600": "#f80e95",
  "700": "#d90880",
  "800": "#b3086b",
  "900": "#8e0855",
  "DEFAULT": "#FF79C6",
}
`;

exports[`shades should generate shades from color correctly 9`] = `
Object {
  "100": "#e9dafe",
  "200": "#d3b6fc",
  "300": "#bd93f9",
  "400": "#a771f6",
  "50": "#fefeff",
  "500": "#924ff2",
  "600": "#7d2eed",
  "700": "#6916e0",
  "800": "#5914bb",
  "900": "#491298",
  "DEFAULT": "#BD93F9",
}
`;

exports[`shades should generate shades from color correctly 10`] = `
Object {
  "100": "#ffc6c6",
  "200": "#ffa0a0",
  "300": "#ff7b7b",
  "400": "#ff5555",
  "50": "#ffecec",
  "500": "#fd3131",
  "600": "#fb0e0e",
  "700": "#dd0606",
  "800": "#b70707",
  "900": "#910707",
  "DEFAULT": "#FF5555",
}
`;

exports[`shades should generate shades from color correctly 11`] = `
Object {
  "100": "#fafed3",
  "200": "#f6fcaf",
  "300": "#f1fa8c",
  "400": "#ebf769",
  "50": "#fefff8",
  "500": "#e5f347",
  "600": "#dfef26",
  "700": "#ccdd12",
  "800": "#abb811",
  "900": "#899410",
  "DEFAULT": "#F1FA8C",
}
`;
